File: Packages/HistoryKit/Tests/HistoryKitTests/HistoryIndexTests.swift
import TestKit
import Foundation
@testable import HistoryKit

/// Fixed whole-second "now" so JSON round-trips compare equal under ISO-8601.
private let now = Date(timeIntervalSince1970: 1_900_000_000)

private func entry(_ kind: HistoryKind = .screenshot, daysAgo: Double = 0) -> HistoryEntry {
    HistoryEntry(kind: kind, date: now.addingTimeInterval(-daysAgo * 86_400),
                 imageFile: kind == .screenshot ? "img.png" : nil,
                 filePath: kind == .recording ? "/tmp/rec.mp4" : nil,
                 thumbFile: "thumb.jpg")
}

let historyIndexTests: [TestCase] = [
    TestCase("addingInsertsNewestFirst") { t in
        let a = entry(daysAgo: 1), b = entry()
        var idx = HistoryIndex()
        idx = idx.adding(a, cap: 50, now: now).index
        let (idx2, evicted) = idx.adding(b, cap: 50, now: now)
        t.equal(idx2.entries.map(\.id), [b.id, a.id])
        t.isTrue(evicted.isEmpty)
    },
    TestCase("countCapEvictsOldest") { t in
        let a = entry(daysAgo: 2), b = entry(daysAgo: 1), c = entry()
        var idx = HistoryIndex()
        idx = idx.adding(a, cap: 2, now: now).index
        idx = idx.adding(b, cap: 2, now: now).index
        let (idx2, evicted) = idx.adding(c, cap: 2, now: now)
        t.equal(idx2.entries.map(\.id), [c.id, b.id])
        t.equal(evicted.map(\.id), [a.id])
    },
    TestCase("entriesOlderThan30DaysArePruned") { t in
        let old = entry(daysAgo: 31), fresh = entry()
        let idx = HistoryIndex(entries: [fresh, old])
        let (pruned, evicted) = idx.adding(entry(daysAgo: 0.5), cap: 50, now: now)
        t.equal(pruned.entries.count, 2)
        t.equal(evicted.map(\.id), [old.id])
    },
    TestCase("exactly30DayOldEntrySurvives") { t in
        let edge = entry(daysAgo: 30)
        let (idx, evicted) = HistoryIndex(entries: [edge]).pruned(cap: 50, now: now)
        t.equal(idx.entries.map(\.id), [edge.id])
        t.isTrue(evicted.isEmpty)
    },
    TestCase("prunedAppliesCapAndAgeAtLoad") { t in
        let a = entry(), b = entry(daysAgo: 1), old = entry(daysAgo: 40)
        let (idx, evicted) = HistoryIndex(entries: [a, b, old]).pruned(cap: 1, now: now)
        t.equal(idx.entries.map(\.id), [a.id])
        t.equal(Set(evicted.map(\.id)), Set([b.id, old.id]))
    },
    TestCase("removingReturnsEntry") { t in
        let a = entry(), b = entry(daysAgo: 1)
        let (idx, removed) = HistoryIndex(entries: [a, b]).removing(id: b.id)
        t.equal(idx.entries.map(\.id), [a.id])
        t.equal(removed?.id, b.id)
    },
    TestCase("removingUnknownIDIsNoOp") { t in
        let a = entry()
        let (idx, removed) = HistoryIndex(entries: [a]).removing(id: UUID())
        t.equal(idx.entries.map(\.id), [a.id])
        t.isNil(removed)
    },
    TestCase("prunedOfMissingFilesDropsOnlyMissing") { t in
        let alive = entry(.recording), dead = entry(.recording, daysAgo: 1)
        let (idx, removed) = HistoryIndex(entries: [alive, dead])
            .prunedOfMissingFiles { $0.id == alive.id }
        t.equal(idx.entries.map(\.id), [alive.id])
        t.equal(removed.map(\.id), [dead.id])
    },
    TestCase("jsonRoundTrip") { t in
        let idx = HistoryIndex(entries: [entry(), entry(.recording, daysAgo: 1)])
        guard let data = t.unwrap(try? idx.jsonData()) else { return }
        guard let back = t.unwrap(try? HistoryIndex(jsonData: data)) else { return }
        t.equal(back, idx)
    },
    TestCase("corruptJSONThrows") { t in
        t.isNil(try? HistoryIndex(jsonData: Data("not json".utf8)))
    },
]

File: windows/tests/BetterScreenshot.Tests/HistoryIndexTests.cs
using System.Text.Json;
using BetterScreenshot.History;
using Xunit;

namespace BetterScreenshot.Tests;

public class HistoryIndexTests
{
    private static HistoryEntry Shot(DateTime date, Guid? id = null) => new()
    {
        Id = id ?? Guid.NewGuid(),
        Kind = HistoryKind.Screenshot,
        Date = date,
        ImageFile = "a.png",
        ThumbFile = "a-thumb.jpg",
    };

    private static readonly DateTime Now = new(2026, 6, 1, 12, 0, 0, DateTimeKind.Utc);

    [Fact]
    public void AddingInsertsNewestFirst()
    {
        var idx = new HistoryIndex();
        var (i1, _) = idx.Adding(Shot(Now.AddMinutes(-2)), cap: 10, now: Now);
        var newest = Shot(Now);
        var (i2, _) = i1.Adding(newest, cap: 10, now: Now);
        Assert.Equal(newest.Id, i2.Entries[0].Id);
        Assert.Equal(2, i2.Entries.Count);
    }

    [Fact]
    public void CountCapEvictsOldest()
    {
        // Entries are newest-FIRST by insertion order; the first one added (oldest) is evicted at the cap.
        var idx = new HistoryIndex();
        HistoryEntry? oldest = null;
        for (int k = 0; k < 4; k++)
        {
            var e = Shot(Now.AddMinutes(-(3 - k))); // k=0 oldest date, added first
            if (k == 0) oldest = e;
            (idx, _) = idx.Adding(e, cap: 3, now: Now);
        }
        Assert.Equal(3, idx.Entries.Count);
        Assert.DoesNotContain(idx.Entries, e => e.Id == oldest!.Id);
    }

    [Fact]
    public void EntriesOlderThan30DaysArePruned()
    {
        var idx = new HistoryIndex(new[] { Shot(Now.AddDays(-31)) });
        var (pruned, evicted) = idx.Pruned(cap: 100, now: Now);
        Assert.Empty(pruned.Entries);
        Assert.Single(evicted);
    }

    [Fact]
    public void Exactly30DayOldEntrySurvives()
    {
        var idx = new HistoryIndex(new[] { Shot(Now.AddDays(-30)) });
        var (pruned, _) = idx.Pruned(cap: 100, now: Now);
        Assert.Single(pruned.Entries);
    }

    [Fact]
    public void RemovingReturnsEntryAndUnknownIsNoOp()
    {
        var e = Shot(Now);
        var idx = new HistoryIndex(new[] { e });
        var (after, removed) = idx.Removing(e.Id);
        Assert.Equal(e.Id, removed!.Id);
        Assert.Empty(after.Entries);

        var (same, none) = idx.Removing(Guid.NewGuid());
        Assert.Null(none);
        Assert.Single(same.Entries);
    }

    [Fact]
    public void PrunedOfMissingFilesDropsOnlyMissing()
    {
        var present = Shot(Now, Guid.NewGuid());
        var missing = Shot(Now, Guid.NewGuid());
        var idx = new HistoryIndex(new[] { present, missing });
        var (kept, removed) = idx.PrunedOfMissingFiles(e => e.Id == present.Id);
        Assert.Single(kept.Entries);
        Assert.Equal(present.Id, kept.Entries[0].Id);
        Assert.Single(removed);
        Assert.Equal(missing.Id, removed[0].Id);
    }

    [Fact]
    public void JsonRoundTrip()
    {
        var idx = new HistoryIndex(new[] { Shot(Now), Shot(Now.AddMinutes(-5)) });
        var round = HistoryIndex.FromJson(idx.ToJson());
        Assert.Equal(idx, round);
    }

    [Fact]
    public void CorruptJsonThrows()
    {
        Assert.ThrowsAny<JsonException>(() => HistoryIndex.FromJson("{ not json ]"));
    }
}
